src/main/scala/io/iohk/ethereum/network/discovery/Secp256k1SigAlg.scala

```diff
@@ -0,0 +1,180 @@
+package io.iohk.ethereum.network.discovery
+
+import akka.util.ByteString
+import io.iohk.ethereum.crypto
+import io.iohk.ethereum.crypto.ECDSASignature
+import io.iohk.scalanet.discovery.crypto.{SigAlg, PublicKey, PrivateKey, Signature}
+import io.iohk.ethereum.nodebuilder.SecureRandomBuilder
+import scodec.bits.BitVector
+import scodec.{Attempt, Err}
+import scodec.bits.BitVector
+
+object Secp256k1SigAlg extends SigAlg with SecureRandomBuilder {
+  override val name = "secp256k1"
+
+  override val PrivateKeyBytesSize = 32
+
+  // A Secp256k1 public key is 32 bytes compressed or 64 bytes uncompressed,
+  // with a 1 byte prefix showing which version it is.
+  // See https://davidederosa.com/basic-blockchain-programming/elliptic-curve-keys
+  //
+  // However in the discovery v4 protocol the prefix is omitted.
+  override val PublicKeyBytesSize = 64
+
+  // A normal Secp256k1 signature consists of 2 bigints `r` and `s` followed by a recovery ID `v`,
+  // but it can be just 64 bytes if that's omitted, like in the ENR.
+  override val SignatureBytesSize = 65
+
+  val SignatureWithoutRecoveryBytesSize = 64
+  val PublicKeyCompressedBytesSize = 33
+
+  override def newKeyPair: (PublicKey, PrivateKey) = {
+    val keyPair = crypto.generateKeyPair(secureRandom)
+    val (privateKeyBytes, publicKeyBytes) = crypto.keyPairToByteArrays(keyPair)
+
+    val publicKey = toPublicKey(publicKeyBytes)
+    val privateKey = toPrivateKey(privateKeyBytes)
+
+    publicKey -> privateKey
+  }
+
+  override def sign(privateKey: PrivateKey, data: BitVector): Signature = {
+    val message = crypto.kec256(data.toByteArray)
+    val keyPair = crypto.keyPairFromPrvKey(privateKey.toByteArray)
+    val sig = ECDSASignature.sign(message, keyPair)
+    toSignature(sig)
+  }
+
+  // ENR wants the signature without recovery ID, just 64 bytes.
+  // The Packet on the other hand has the full 65 bytes.
+  override def removeRecoveryId(signature: Signature): Signature = {
+    signature.size / 8 match {
+      case SignatureBytesSize =>
+        Signature(signature.dropRight(8))
+      case SignatureWithoutRecoveryBytesSize =>
+        signature
+      case other =>
+        throw new IllegalArgumentException(s"Unexpected signature size: $other bytes")
+    }
+  }
+
+  override def compressPublicKey(publicKey: PublicKey): PublicKey = {
+    publicKey.size / 8 match {
+      case PublicKeyBytesSize =>
+        // This is a public key without the prefix, it consists of an x and y bigint.
+        // To compress we drop y, and the first byte becomes 02 for even values of y and 03 for odd values.
+        val (xbs, ybs) = publicKey.splitAt(publicKey.length / 2)
```

Comment:
compressing can be done by going through bouncy castle only, like:
```
    val params = new ECPublicKeyParameters(
      curve.getCurve.decodePoint(publicKey.toByteArray),
      curve
    )
    params.getQ.getEncoded(true)
```

Reply:
Cool, thanks!

Reply:
Done.